tests/test_filesystems.py

```diff
@@ -52,6 +81,129 @@ async def test_get_directory_duplicate_directory(self, tmp_path):
+            parent_contents, child_contents = setup_test_directory(
+                tmp_src, sub_dir_name
+            )
+            # move file contents to tmp_dst
+            with TemporaryDirectory() as tmp_dst:
+
+                f = LocalFileSystem()
+
+                await f.get_directory(from_path=tmp_src, local_path=tmp_dst)
+                assert set(os.listdir(tmp_dst)) == set(parent_contents)
+                assert set(os.listdir(Path(tmp_dst) / sub_dir_name)) == set(
+                    child_contents
+                )
+
+    async def test_dir_contents_copied_correctly_with_put_directory(self):
+
+        sub_dir_name = "puppy"
+
+        with TemporaryDirectory() as tmp_src:
+            parent_contents, child_contents = setup_test_directory(
+                tmp_src, sub_dir_name
+            )
+            # move file contents to tmp_dst
+            with TemporaryDirectory() as tmp_dst:
+
+                f = LocalFileSystem()
+
+                await f.put_directory(
+                    local_path=tmp_src,
+                    to_path=tmp_dst,
+                )
+
+                assert set(os.listdir(tmp_dst)) == set(parent_contents)
+                assert set(os.listdir(Path(tmp_dst) / sub_dir_name)) == set(
+                    child_contents
+                )
+
+    async def test_dir_contents_copied_correctly_with_put_directory_and_file_pattern(
+        self,
+    ):
+        """Make sure that ignore file behaves properly."""
+
+        sub_dir_name = "puppy"
+
+        with TemporaryDirectory() as tmp_src:
+            parent_contents, child_contents = setup_test_directory(
+                tmp_src, sub_dir_name
+            )
+
+            # ignore .py files
+            ignore_fpath = Path(tmp_src) / ".ignore"
+            with open(ignore_fpath, "w") as f:
+                f.write("*.py")
+
+            # contents without .py files
+            expected_contents = os.listdir(tmp_src)
+
+            # add .py files
+            with open(Path(tmp_src) / "dog.py", "w") as f:
+                f.write("pass")
+
+            with open(Path(tmp_src) / sub_dir_name / "cat.py", "w") as f:
+                f.write("pass")
+
+            # move file contents to tmp_dst
+            with TemporaryDirectory() as tmp_dst:
+
+                f = LocalFileSystem()
+
+                await f.put_directory(
+                    local_path=tmp_src, to_path=tmp_dst, ignore_file=ignore_fpath
+                )
+                assert set(os.listdir(tmp_dst)) == set(expected_contents)
+                assert set(os.listdir(Path(tmp_dst) / sub_dir_name)) == set(
+                    child_contents
+                )
+
+    async def test_dir_contents_copied_correctly_with_put_directory_and_directory_pattern(
+        self,
+    ):
+        """Make sure that ignore file behaves properly."""
+
+        sub_dir_name = "puppy"
+        skip_sub_dir = "kitty"
+
+        with TemporaryDirectory() as tmp_src:
```

Comment:
In the future, I'd use the `tmp_path` fixture and just make two directories in it for a source and destination — no need for all this nesting

Reply:
Then your `setup_test_directory` could just be a fixture that uses `tmp_path`.

Reply:
Awesome - I'll refactor this once I'm back from break.